### DISCUSSION TITLE
Annotating the decorator with arguments that works for both sync/async functions

### DISCUSSION THREAD

**Discussion by Galtozzy** [python/typing#1847] [Q&A] [answered]
Hey there!

For context - I've tried asking the same question on typing Gitter and had no luck there since I believe the visibility is quite low and I got no comments to the message up to now.
I've checked the issues [1](https://github.com/microsoft/pyright/issues/2142), [2](https://github.com/microsoft/pyright/issues/5014)

I am taking my luck here just to understand if what I'm trying to achieve is possible at all or if I shouldn't bother trying.

The idea seems simple - we need to preserve all types of a function the decorator with arguments is being applied to. However, the issue comes when the decorator is supposed to work with both sync and async functions.

The solution I've come close with is the following:
```
R = TypeVar('R')
P = ParamSpec('P')


def is_coroutine(func: Union[Callable[P, R], Callable[P, Awaitable[R]]]) -> TypeGuard[Callable[P, Awaitable[R]]]:
    return inspect.iscoroutinefunction(func)


def my_dec(
    param1: str, param2: Union[int, None] = None
) -> Callable[[Union[Callable[P, Awaitable[R]], Callable[P, R]]], Union[Callable[P, Awaitable[R]], Callable[P, R]]]:

    def decorator(_func: Union[Callable[P, R], Callable[P, Awaitable[R]]]) -> Union[Callable[P, R], Callable[P, Awaitable[R]]]:
        if is_coroutine(_func):
            _awaitable_func = _func

            @wraps(_awaitable_func)
            async def _async_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
                return await _awaitable_func(*args, **kwargs)
            return _async_wrapper

        else:

            @wraps(_func)
            def _sync_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
                return cast(R, _func(*args, **kwargs))

            return cast(Callable[P, R], _sync_wrapper)

    return decorator
    
    
@my_dec(param1='test')
def test() -> str:
    return 'test return'
    
    
return_value: str = test()
```
However, `TypeGuard` doesn't narrow down the type thus mypy throws an error on `return_value: str = test()`:
`Incompatible types in assignment (expression has type "Awaitable[str] | str", variable has type "str")  [assignment]`

I have also tried `@typing.overload` uproach but it seems it doesn't work in this case since overloading happens inside the `my_dec`.
I would appreciate any help, directions and resources I could look into to solve this.

**Comment by erictraut** [comment, marked as answer]
Here's a potential solution. It type checks without errors in pyright. Mypy produces a few overload-related errors that appear to be false positives, so you could add `# type: ignore` comments to suppress them.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCA7gIZIwkBGANgKYA0UAwiddVXYwArgxgDGYaozAA3WiGpgSAEwBQoSLESo0AfVoAPGLRQBnJGH2ZsufABVEtAJJ6F4aKj0Ja-fFhx5MeweACuMKi0wH4oboYo9kohYXxCeiae%2BEQgJAh2cjLB3mq%2BYAFBANoAVMVcjABKALoAFHJQDVAx-ABczKzsNLSF5VAVUAA%2BhKTkHN3VVfRyAJRQALQAfFCWLraFLGxjPYzEZBRdhRNVLfWNILQwfiAo3nkFKMGh4UY1zdNycvzsegkAynBhADyIAIegB-BqPDAfEE1GmJ0aUAAAmIJFJZKcGllgFA1LkOniSmVKrVMYioHpaNRgIw1M02htOnRtsM9lsJmTZot2psDr1dqMDhMEeSGgA6CUfREo8SSaTyRHY3H4tiE0q9ao1SnU2n0nlM7oaqpcpaMrZGkWiiViqWNJV4-gEtREo11UUNbU03F6s18nYjfbMiaDfXmkmc%2Bamjph1mCoNVKoh33Mi1kxHWj5KiBwNRZCEIEipCAARjaehgIEYBaLACY2qh8EMAHJGWhQAC8UBbDxNUH%2BQJBYLCluRqLlGMV2TzuBIfBALpJbvddKeDOjftjgfGCamiN7ycN-rZQoTI-TkrJMrR8rJSunqTnC76pPd3tXoY31V3jX365TJLPRoM1vKdXBnR91UXNNGhXYcP3-Td2QTJM-0PZ9vwaX9eQQgUt0OZChgPFlqkAhokBxJA9FyXB8kCB4alg-h4Wg0U1BIAMxkYjs3zCW1XwaJEUjSPQGPY486EY95%2BMaEgh34KB7Vk8E1CEhAXBAGpikLNA9DaLgxW0vRGFKABrUgQB0vSxTMwzewqUj%2BPOS5rigMTyFxNyt0YzTDOM4obIsvR3hY8knKuG42LklTUjUiQ%2BPJKlKROELpVUkTJJSu1sjUKLVPUnzAqs3yoFM8zLKgfSAp0uyHNfMKXO8rTAr8qqgviuqLnC3Fcpi9T4vqm571nXAPiRbNc1cGpqxIEt2wAIh0cs5veJVFpgHLwRqXtyxAEcBqgBbaHLKABrmj5RA2sJuLWy6IWCsaczzKbCxm4t5rW5a5CUq7VqO9bvruyMKQrPbOpcw7jtO87IvBa6-phsItrkIA)

```python
from typing import Awaitable, Callable, Protocol, overload
from typing_extensions import TypeIs
from inspect import iscoroutinefunction
from functools import wraps

def is_coroutine[**P, R](
    func: Callable[P, R | Awaitable[R]],
) -> TypeIs[Callable[P, Awaitable[R]]]:
    return iscoroutinefunction(func)

class SyncOrAsync(Protocol):
    @overload
    def __call__[**P, R](
        self, _func: Callable[P, Awaitable[R]]
    ) -> Callable[P, Awaitable[R]]:
        ...

    @overload
    def __call__[**P, R](self, _func: Callable[P, R]) -> Callable[P, R]:
        ...

    def __call__[**P, R](
        self, _func: Callable[P, Awaitable[R]] | Callable[P, R]
    ) -> Callable[P, Awaitable[R]] | Callable[P, R]:
        ...

def my_dec(param1: str, param2: int | None = None) -> SyncOrAsync:
    @overload
    def decorator[**P, R](
        _func: Callable[P, Awaitable[R]],
    ) -> Callable[P, Awaitable[R]]:
        ...

    @overload
    def decorator[**P, R](
        _func: Callable[P, R],
    ) -> Callable[P, R]:
        ...

    def decorator[**P, R](
        _func: Callable[P, Awaitable[R]] | Callable[P, R],
    ) -> Callable[P, Awaitable[R]] | Callable[P, R]:
        if is_coroutine(_func):
            _awaitable_func = _func

            @wraps(_awaitable_func)
            async def _async_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
                return await _awaitable_func(*args, **kwargs)

            return _async_wrapper

        else:

            @wraps(_func)
            def _sync_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
                return _func(*args, **kwargs)

            return _sync_wrapper

    return decorator

@my_dec(param1="test")
def test_sync() -> str:
    return "test return"

v_sync = test_sync()

@my_dec(param1="test")
async def test_async() -> str:
    return "test return"

v_async = test_async()

```

**Reply by Galtozzy** [reply]
Worked like a charm, thank you very much!

**Reply by staciax** [reply]
thank you so much <3
